### ISSUE TITLE
Compilation error using boost 1.80 beta 1 with Visual Studio 2017 and EditAndContinue

### ISSUE THREAD

**Issue by thebrandre** [chriskohlhoff/asio#1089]
Even though this is rather a compiler bug than a bug in asio, it would be nice to avoid breaking support for Visual Studio 2017's compiler.

Here is a [minimal example on godbolt](https://godbolt.org/z/a1zvPnse3) where commit f7faa79cfc6952a13e457cd0577d018debeb7a22 breaks compilation with Visual Studio 2017's compiler (msvc 19.16) when the debug-information format [EditAndContinue /ZI](https://docs.microsoft.com/en-us/cpp/build/reference/z7-zi-zi-debug-information-format?view=msvc-170) is used.

The error message is:

> error C2131: expression did not evaluate to a constant
> note: failure was caused by non-constant arguments or reference to a non-constant symbol
>note: see usage of '__LINE__Var'

It can be fixed by using `const` instead of `constexpr` for `BOOST_STATIC_CONSTEXPR` (if compiled by the native Visual Studio 2017 compiler) here:
```cpp
# define BOOST_ASIO_ERROR_LOCATION(e) \
  do { \
    BOOST_STATIC_CONSTEXPR boost::source_location loc \
      = BOOST_CURRENT_LOCATION; \
    (e).assign((e), &loc); \
  } while (false)
```

Note: As demonstated in the godbolt link, it works fine with the more recent compilers of Visual Studio 2019 and 2022.

**Comment by pdimov** [comment]
Should be fixed by https://github.com/boostorg/assert/commit/3d6e55bddf6f12dbb2bf0ef634f1b145990854d3. I'll try to get this merged for 1.80.

**Comment by thebrandre** [comment]
Thank you!

I applied your patch to the beta release and now our 6 million line of C++ code with heavy boost usage compiles again. :-)
